allow user passing relative paths in include_dirs within setuptools.setup

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38264 allow user passing relative paths in include_dirs within setuptools.setup**

This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

Commit:
allow user passing relative paths in include_dirs within setuptools.setup

[ghstack-poisoned]

zou3519 commented on torch/utils/cpp_extension.py:
```diff
@@ -407,6 +407,18 @@ def unix_wrap_ninja_compile(sources,
             # (`objects`) get generated with absolute paths.
             output_dir = os.path.abspath(output_dir)
+
+            # Convert relative path in self.compilr.include_dirs to absolute path if any,
```
nit: "self.compiler.include-dirs"

zou3519 commented on torch/utils/cpp_extension.py:
```diff
@@ -407,6 +407,18 @@ def unix_wrap_ninja_compile(sources,
             output_dir = os.path.abspath(output_dir)
+
+            # Convert relative path in self.compilr.include_dirs to absolute path if any,
+            # For ninja build, the build location is not local, the build happns
```
nit: happns -> happens

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]

zou3519 commented on torch/utils/cpp_extension.py:
```diff
@@ -407,6 +407,18 @@ def unix_wrap_ninja_compile(sources,
             # Use absolute path for output_dir so that the object file paths
             # (`objects`) get generated with absolute paths.
             output_dir = os.path.abspath(output_dir)
+
```
How difficult is it to add a test to https://github.com/pytorch/pytorch/blob/master/test/test_cpp_extensions_aot.py? I don't think we need to test the library_dir case because that seems difficult to test, but I feel like we should at least test that `include_dir` is working correctly.

glaringlee replied:
@zou3519. 
I think we don't need to add this into auto test script. Since this involves user action as well, the build script (setup.py) is created manually by the user who want to build cpp extension. and different people might have different build scripts based on their files name and location. We'd better test it when doing the code change and rely on people reporting issues if any in this case. It is not like pytorch, we control the build script.

zou3519 replied:
The thing I'd like to test here is that "relative include_dirs work". We don't have to check that "arbitrary user configurations work".

I think this is possible, because we test cpp extensions over a custom build script: https://github.com/pytorch/pytorch/tree/master/test/cpp_extensions. We should be able to test that "relative include_dirs work" by adding an includes folder that contains some code to that directory and then adding an "include_dirs" option to the setup.py: https://github.com/pytorch/pytorch/blob/ba0851326c87256a819197c133b44f3e617b0608/test/cpp_extensions/setup.py#L19-L29 .

The problem with relying on people to report issues is that there is the risk that we release a version of PyTorch and break some users' code; that's what happened in this case.

glaringlee commented on torch/utils/cpp_extension.py:
```diff
@@ -407,6 +407,18 @@ def unix_wrap_ninja_compile(sources,
+                for i in range(len(self_compiler_include_dirs)):
+                    if not os.path.isabs(self_compiler_include_dirs[i]):
+                        self_compiler_include_dirs[i] = os.path.abspath(self_compiler_include_dirs[i])
+
```
@zou3519 , this is not the same include_dirs.
The include_dirs that fixed in this PR is located here:
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```
while the include_dirs from kwargs is located here:
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu'],
             include_dirs=['/home/abc/def/includes'] -> absolute path only
        )
    ],
    cmdclass={
        'build_ext': BuildExtension
  })

The kwargs one support absolute path only and it is as designed. The include_dirs that I want to keep consistent with jit extension is the one outside CPP/CUDAExtension, and this is the one that user complained.

zou3519 replied:
Oh, I see, I didn't realize. Thanks for the clarification.

glaringlee commented on torch/utils/cpp_extension.py:
```diff
@@ -407,6 +407,18 @@ def unix_wrap_ninja_compile(sources,
             # Use absolute path for output_dir so that the object file paths
             # (`objects`) get generated with absolute paths.
             output_dir = os.path.abspath(output_dir)
+
```
See my comment above. We don't need to change anything within CUDA/CPPExtension.

zou3519 replied:
Thanks for the clarification!

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]

zou3519 commented:
Hmm, it looks like the fix wasn't enough for windows? (See windows tests)

glaringlee commented:
@zou3519 yep, I don't have env for windows, not aware of it. I am working on the fix, will update this later.

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]

ezyang commented on test/cpp_extensions/extension.cpp:
```diff
@@ -1,5 +1,8 @@
 #include <torch/extension.h>
 
```
This should be `include_dirs`

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1724,3 +1740,10 @@ def _is_cuda_file(path):
         valid_ext.append('.hip')
     return os.path.splitext(path)[1] in valid_ext
+
+def _to_absolute_path(paths):
```
The name here implies that this is a pure function, but actually it mutates the list in place. A more descriptive name would be better.

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -1724,3 +1740,10 @@ def _is_cuda_file(path):
+
+def _to_absolute_path(paths):
+    if not paths:
+        return
```
This special case is totally unnecessary since the loop will quit out if the paths array is empty

glaringlee replied:
This is to prevent calling len() on None which will result in a crash.

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -536,6 +544,14 @@ def win_wrap_ninja_compile(sources,
+            # For ninja build, the build location is not local, the build happens
+            # in a in script created build folder, relative path lost their correctness.
+            # To be consistent with jit extension, we allow user to enter relative include_dirs
+            # in setuptools.setup, and we convert the relative path to absolute path here
```
Instead of duplicating this comment in two places (and then forcing subsequent editors to update it in both places), convert it into a note! In one location, say `Note [Absolute include_dirs]` and then the rest of the note, and then in relevant locations just say `See Note [Absolute include_dirs]`. You should also add a reference to this note at `_to_absolute_path`

ezyang commented:
I'm actually not sure why I was tagged on this PR; @zou3519 seems like the right final reviewer, I'll let him make the final call

glaringlee commented:
@ezyang It seems github added all the potential owner of this file into the code review automatically, I will let @zou3519 do the rest of review.

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]

zou3519 reviewed (approved):
LGTM, I forgot that the windows ninja ahead-of-time c++ extensions goes through another codepath

zou3519 commented on torch/utils/cpp_extension.py:
```diff
@@ -359,6 +359,14 @@ def unix_cuda_flags(cflags):
 
+        def convert_to_absolute_paths_inplace(paths):
+            # Helper function. See Note [Absolute include_dirs]
+            if not paths:
```
nit: It's better as `if paths is not None` to make it clear that we're testing for non-None-ness, otherwise this might look like a check that paths is an empty list (which is unnecessary)

Commit:
Update on "allow user passing relative paths in include_dirs within setuptools.setup"


This is to fix #37707

Tested with a similar setup as below (provided by user in #37707) with both relative path and absolute path. All works fine.
```
setup(
    name='lltm_cuda',
    ext_modules=[
        CUDAExtension('lltm_cuda', [
            'lltm_cuda.cpp',
            'lltm_cuda_kernel.cu']
        ),
       include_dirs=['includes']
    ],
    cmdclass={
        'build_ext': BuildExtension
  })
```

Differential Revision: [D21509277](https://our.internmc.facebook.com/intern/diff/D21509277)

[ghstack-poisoned]
